File: ios/TextClassifier.swift
import Foundation
import OSLog
import NitroModules

class TextClassifier: HybridTextClassifierSpec {

    private var isInitialized: Bool = false

    func initialize(source: ModelSource, options: TextClassifierOptions?) throws -> Promise<Void> {
        return Promise.async {
            // Simulate initialization with a small delay
            try await Task.sleep(nanoseconds: 100_000_000) // 0.1 seconds
            self.isInitialized = true
            
        }
    }

    func unload() throws -> Void {
        isInitialized = false
        
    }

    func classify(text: String) throws -> Promise<TextClassifierResult> {
        return Promise.async {
            guard self.isInitialized else {
                throw NSError(domain: "TextClassifier", code: -1, userInfo: [NSLocalizedDescriptionKey: "TextClassifier is not initialized. Call initialize() first."])
            }

            // Return hardcoded classification results for testing
            let categories: [Category] = [
                Category(
                    index: 0,
                    score: 0.85,
                    displayName: "Positive",
                    categoryName: "positive"
                ),
                Category(
                    index: 1,
                    score: 0.15,
                    displayName: "Negative",
                    categoryName: "negative"
                )
            ]

            
            
            return TextClassifierResult(classifications: categories)
        }
    }
}

File: android/src/main/java/com/margelo/nitro/aistack/TextClassifier.kt
package com.margelo.nitro.aistack

import android.util.Log
import com.facebook.proguard.annotations.DoNotStrip
import com.google.mediapipe.tasks.core.BaseOptions
import com.google.mediapipe.tasks.text.textclassifier.TextClassifier
import com.margelo.nitro.core.Promise
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

@DoNotStrip
class TextClassifier : HybridTextClassifierSpec() {

  private val moduleBase =
    MediaPipeModuleBase<TextClassifier, TextClassifierOptions>("NitroTextClassifier")

  override fun initialize(source: ModelSource, options: TextClassifierOptions?): Promise<Unit> {
    return moduleBase.initializeWithCaching(source, options) { modelPath, opts ->
      val baseOptions = BaseOptions.builder()
        .setModelAssetPath(modelPath)
        .build()

      val classifierOptions = TextClassifier.TextClassifierOptions.builder()
        .setBaseOptions(baseOptions)
        .apply {
          opts?.maxResults?.let { maxResults ->
            setMaxResults(maxResults.toInt())
          }
          opts?.scoreThreshold?.let { scoreThreshold ->
            setScoreThreshold(scoreThreshold.toFloat())
          }
          opts?.categoryAllowlist?.let { setCategoryAllowlist(it.toList()) }
          opts?.categoryDenylist?.let { setCategoryDenylist(it.toList()) }
          opts?.displayNamesLocale?.let { setDisplayNamesLocale(it) }
        }
        .build()

      TextClassifier.createFromOptions(moduleBase.reactContext, classifierOptions)
    }
  }

  override fun classify(text: String): Promise<TextClassifierResult> {
    return Promise.async {
      try {
        val classifier = moduleBase.requireModule()

        val results = withContext(Dispatchers.Default) {
          classifier.classify(text)
        }

        val classifications =
          results.classificationResult().classifications().flatMap { classification ->
            classification.categories().map { category ->
              Category(
                score = category.score().toDouble(),
                index = category.index().toDouble(),
                categoryName = category.categoryName() ?: "",
                displayName = category.displayName() ?: ""
              )
            }
          }.toTypedArray()

        Log.d(
          moduleBase.tag,
          "Classification completed for text: ${text.take(50)}, results: ${classifications.size}"
        )
        TextClassifierResult(classifications = classifications)
      } catch (e: RuntimeException) {
        throw IllegalStateException("Classification failed: ${e.message}", e)
      }
    }
  }

  override fun unload() {
    moduleBase.unload()
  }
}
